intermediate_source/rpc_tutorial.rst

```diff
@@ -0,0 +1,529 @@
+            # get RRefs of decoder
+            remote_params.extend(_remote_method(_parameter_rrefs, self.decoder_rref))
+            return remote_params
+
+Now, we are ready to implement the training loop. After initializing the model
+arguments, we create the ``RNNModel`` and the ``DistributedOptimizer``. The
+distributed optimizer will take a list of parameter ``RRefs``, find all distinct
+owner workers, and create the given local optimizer (i.e., ``SGD`` in this case)
+on each of the owner worker using the given arguments (i.e., ``lr=0.05``).
+
+In the training loop, it first creates a distributed autograd context, which
+will help the distributed autograd engine to find gradients and involved RPC
+send/recv functions. Then, it kicks off the forward pass as if it is a local
+model, and run the distributed backward pass. For the distributed backward, you
+only need to specify a list of roots, in this case, it is the loss ``Tensor``.
+The distributed autograd engine will traverse the distributed graph
+automatically and write gradients properly. Next, it runs the ``step``
+API on the distributed optimizer, which will reach out to all involved local
+optimizers to update model parameters. Compared to local training, one minor
+difference is that you don't need to run ``zero_grad()`` because each autograd
+context has dedicated space to store gradients, and as we create a context
+per iteration, those gradients from different iterations will not accumulate to
+the same set of ``Tensors``.
+
+.. code:: python
+
+    def run_trainer():
+        batch = 5
+        ntoken = 10
+        ninp = 2
+
+        nhid = 3
+        nindices = 3
+        nlayers = 4
+        hidden = (
+            torch.randn(nlayers, nindices, nhid),
+            torch.randn(nlayers, nindices, nhid)
+        )
+
+        model = rnn.RNNModel('ps', ntoken, ninp, nhid, nlayers)
+
+        # setup distributed optimizer
+        opt = DistributedOptimizer(
+            optim.SGD,
+            model.parameter_rrefs(),
+            lr=0.05,
+        )
+
+        # train for 10 iterations
+        for epoch in range(10):
+            # create distributed autograd context
+            with dist_autograd.context():
+                inp = torch.LongTensor(batch, nindices) % ntoken
+                hidden[0].detach_()
+                hidden[1].detach_()
+                output, hidden = model(inp, hidden)
+                # run distributed backward pass
+                dist_autograd.backward([output.sum()])
+                # run distributed optimizer
+                opt.step()
+                # not necessary to zero grads as each iteration creates a different
+                # distributed autograd context which hosts different grads
+                print("Training epoch {}".format(epoch))
+
+
+Finally, let's add some glue code to launch the parameter server and the trainer
+processes.
+
+
+.. code:: python
+
+    def run_ps():
```

Comment:
Don't we need to `rpc.init_rpc` on the node that hosts the param server?

Reply:
`init_rpc` is in `run_worker` which is the entry point for this example. I see how this can be confusing, as I used different ways to spawn RPC processes. let me change this to let two examples using similar entry functions.